pysat/utils/registry.py

```diff
@@ -15,117 +15,296 @@
+            # setup `of dict` part of dict of dicts
+            pysat.user_modules[platform] = {}
+        # only register name if not present under platform
+        if name not in pysat.user_modules[platform]:
+            logger.info('Registering user module {}'.format(module_name))
+            # add to current user modules structure
+            pysat.user_modules[platform][name] = module_name
+            # store user modules to disk
+            store()
+        else:
+            # platform/name combination already registered
+            estr = ' '.join(('An instrument has already been registered using',
+                             'platform:', platform, 'and name:', name,
+                             'which maps to:', module_name))
+            raise ValueError(estr)
 
+    return
 
-    testInst = Instrument()
 
+def register_by_module(module):
+    """Register all sub-modules attached to input module
+
+    Enables instantiation of a third-party Instrument
+    module using
+    ::
+        inst = pysat.Instrument(platform, name)
+
+    Parameters
+    ----------
+    module : Python module
+        Module with one or more pysat.Instrument support modules
+        attached as sub-modules to the input `module`
+
+    Raises
+    ------
+    ValueError
+        If platform and name associated with a module are already registered
 
+    Note
+    ----
+    Gets a list of sub-modules by using the __all__ attribute,
+    defined in the module's __init__.py
+
+    Examples
+    --------
+    ::
+        import pysat
+        import pysatModels
+        pysat.utils.registry.register_by_module(pysatModels.models)
+
+        import pysatSpaceWeather
+        pysat.utils.registry.register_by_module(pysatSpaceWeather.instruments)
 
     """
-    import importlib
-    inst_module = importlib.import_module(module_name)
 
-    if module_name not in pysat.user_modules:
-        print('registering user module {}'.format(module_name))
-        pysat.user_modules.append(module_name)
-        store()
+    # get a list of all user specified modules attached to module
+    module_names = module.__all__
+    # add in package preamble
+    module_names = [module.__name__ + '.' + mod for mod in module_names]
+    # register all of them
+    register(module_names)
 
-    return inst_module
+    return
 
 
-def remove(*module_names):
+def remove(platforms, names=None):
     """Removes module from registered user modules
 
     Parameters
-    -----------
-    module_names : str or multiple str
-        full module package names to remove from registry
-
-    Notes
+    ----------
+    platforms : str of list-like of str
+        One or more platform identifiers to remove
+    names : str of list-like of str
+        Name identifiers, paired with platforms, to remove.
+        If names is None, then all instruments under
+        `platforms` will be removed. Supports a mixed
+        combination of name labels and None. (default=None)
```

Comment:
```suggestion
    names : str or list-like of str
        Name identifiers, paired with platforms, to remove.
        If the names element paired with the platform element is None,
        then all instruments under the specified platform will be
        removed. Should be the same type as `platforms`. An empty list
        will default to removing all instruments for all specified
        platforms. (default=[])
```

Reply:
I adopted the suggestion apart from the default.